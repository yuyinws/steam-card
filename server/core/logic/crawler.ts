import { load } from 'cheerio'

export function crawler(html: string) {
  const $ = load(html)

  let gameCount = '0'
  let screenshotCount = '0'
  let artWorkCount = '0'
  let reviewCount = '0'
  let guideCount = '0'
  let badgeCount = '0'

  const groupIconList: string[] = []
  $('.profile_group_links')
    .children()
    .last()
    .children()
    .each((i, el) => {
      const groupIconUrl = $(el).children().first().children().children().attr('src') as string
      groupIconList.unshift(groupIconUrl)
    })

  let groupCount = $('.profile_group_links')
    .children()
    .first()
    .children()
    .find('.profile_count_link_total')
    .text()

  groupCount = groupCount.toString().replaceAll('\n', '').replaceAll('\t', '')

  const badgeIconUrl = $('.favorite_badge_icon').children().attr('src') as string

  $('.count_link_label').each((i, el) => {
    const itemName = $(el).text()
    let count = $(el).next().text()
    count = count.toString().replaceAll('\n', '').replaceAll('\t', '')

    switch (itemName) {
      case 'Games':
        gameCount = count
        break
      case 'Screenshots':
        screenshotCount = count
        break
      case 'Artwork':
        artWorkCount = count
        break
      case 'Reviews':
        reviewCount = count
        break
      case 'Guides':
        guideCount = count
        break
      case 'Badges':
        badgeCount = count
        break
    }
  })

  const playerLevel = $('.persona_name,.persona_level').find('.friendPlayerLevelNum').text()

  const avatarUrl = $('.playerAvatarAutoSizeInner').children().last().attr('src')

  return {
    gameCount,
    groupCount,
    badgeIconUrl,
    groupIconList,
    screenshotCount,
    artWorkCount,
    reviewCount,
    guideCount,
    badgeCount,
    playerLevel,
    avatarUrl,
  }
}
