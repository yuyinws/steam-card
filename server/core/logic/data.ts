import { env } from 'node:process'
import type { PlayedGames, Player } from 'types'

export function data(userInfo: Player, playedGames: PlayedGames) {
  const { avatarfull: avatarUrl, personaname: name, personastate: isOnline } = userInfo
  const blockApps = env.BLOCK_APPS || ''
  const blockAppList = blockApps.split(',')
  // '<' cause svg render error
  const _name = name.replaceAll('<', '&lt;')

  let playTime = 0
  let games = playedGames.games || []

  games.forEach((game) => {
    playTime += game.playtime_2weeks
  })

  playTime = Number.parseInt(String(playTime / 60), 10)
  games = games.filter(game => !blockAppList.includes(String(game.appid)))
  games.splice(5, games.length - 5)

  return { games, playTime, avatarUrl, name: _name, isOnline }
}
