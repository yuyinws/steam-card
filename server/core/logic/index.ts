export * from './crawler'
export * from './data'
export * from './parse'
