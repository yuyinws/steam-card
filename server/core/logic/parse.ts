import type { Config, Statistic } from 'types'
import { defaultStatistics, statisticsList, themes } from '@/utils/constant'

export function parseUrlConfig(_setting: string) {
  const config: Config = {
    theme: 'dark',
    group: false,
    badge: false,
    lang: 'zhCN',
    statistics: [],
    textColor: '',
    bg: '',
  }

  const statisticSet: Set<Statistic> = new Set()

  if (_setting) {
    const settings = _setting.split(',')
    settings.forEach((item: any) => {
      if (themes.includes(item)) {
        config.theme = item
      }
      else if (item.includes('text-')) {
        config.textColor = `#${item.split('-')[1]}`
      }
      else if (item.includes('bg-')) {
        let bg = ''
        if (item.includes('game')) {
          bg = item
        }
        else {
          const colors = item.split('-')
          colors.shift()
          if (colors.length === 1)
            colors.push(colors[0])
          for (let i = 0; i < colors.length; i++) {
            bg += `#${colors[i]}`
            if (i < colors.length - 1)
              bg += ','
          }
        }

        config.bg = bg
      }
      else if (item === 'group') {
        config.group = true
      }
      else if (item === 'badge') {
        config.badge = true
      }
      else if (item === 'zhCN') {
        config.lang = 'zhCN'
      }
      else if (item === 'en') {
        config.lang = 'en'
      }
      else if (statisticsList.includes(item)) {
        statisticSet.add(item)
      }
    })
  }

  defaultStatistics.forEach((item) => {
    statisticSet.add(item)
  })

  const statistics = Array.from(statisticSet)

  if (statistics.length > 3)
    statistics.splice(3, statistics.length - 3)

  config.statistics = statistics
  return {
    config,
  }
}
