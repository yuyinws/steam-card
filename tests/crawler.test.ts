import { readFile } from 'node:fs/promises'
import { join } from 'node:path'
import { cwd } from 'node:process'
import { describe, expect, it } from 'vitest'
import { crawler } from 'server/core/logic/crawler'

describe('crawler', () => {
  it('crawler', async () => {
    const html = await readFile(join(cwd(), 'tests', 'fixtures', 'profile.html'), 'utf-8')

    expect(crawler(html)).toMatchSnapshot()
  })
})
