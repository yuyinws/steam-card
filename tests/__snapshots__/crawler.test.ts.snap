// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`crawler > crawler 1`] = `
{
  "artWorkCount": "3",
  "avatarUrl": "https://avatars.cloudflare.steamstatic.com/53e9999c84e72b247dde54082ca077dca89784d8_full.jpg",
  "badgeCount": "38",
  "badgeIconUrl": "https://community.cloudflare.steamstatic.com/public/images/badges/51_summer2021/54_54.png",
  "gameCount": "218",
  "groupCount": "5",
  "groupIconList": [
    "https://avatars.cloudflare.steamstatic.com/96dbe7106f59729272139f2fa076cddeb7e33730.jpg",
    "https://avatars.cloudflare.steamstatic.com/f2a29784032f714eb79afbd2c94767d17101f871.jpg",
    "https://avatars.cloudflare.steamstatic.com/4bef364761737e84a561c97740064d53d8ba3d36_medium.jpg",
  ],
  "guideCount": "0",
  "playerLevel": "34",
  "reviewCount": "11",
  "screenshotCount": "62",
}
`;
